### PR TITLE
python3Packages.niwidgets: init at 0.2.2

### PR DESCRIPTION
**Pull request by cfhammill**
###### Description of changes

Adds niwidgets and 3 dependencies

python3Packages.niwidgets: init at 0.2.2
python3Packages.pythreejs: init at 2.3.0
python3Packages.ipyvolume: init at 0.5.2
python3Packages.ipywebrtc: init at 0.6.0

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

CC @bcdarwin

### CODE REVIEW

**Comment by bcdarwin** [comment]
@GrahamcOfBorg build python3Packages.niwidgets

**Review by bcdarwin — COMMENTED** [review]
The `pythonImportsCheck` for `niwidgets` appears to be failing; maybe you need to `export HOME=$TMPDIR` or similar for that phase?

**Review by bcdarwin — COMMENTED** [review]
Stylistically, a few more blank lines between logical sections of the expression would be slightly nicer.

File: pkgs/development/python-modules/ipyvolume/default.nix [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  pname = "ipyvolume";
+  version = "0.5.2";
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "Gcve6i2InNG7cPe4L4FIM60LPRFcgXHXt4YIUdj7vSY=";
+  };
+  propagatedBuildInputs = [ ipywidgets pythreejs ipywebrtc pillow ];
+  doCheck = false; # tries to do some headless browsing
+  pythonImportsCheck = [ "ipyvolume" ];
+  meta = {
```
Comment (bcdarwin) [inline pkgs/development/python-modules/ipyvolume/default.nix:20]: consider `meta = with lib; { ... };`

**Comment by bcdarwin** [comment]
Results of `nixpkgs-review pr` on x64 Linux:

```
error: builder for '/nix/store/j4sahn2primmfn8plqy50jc48knnbk63-python3.10-ipyvolume-0.5.2.drv' failed with exit code 1;
       last 10 log lines:
       >   File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
       >   File "<frozen importlib._bootstrap_external>", line 883, in exec_module
       >   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
       >   File "/nix/store/yw8yb6x8iad55hnwr77ch198864qbjrw-python3.10-ipyvolume-0.5.2/lib/python3.10/site-packages/ipyvolume/__init__.py", line 4, in <module>
       >     from . import styles
       >   File "/nix/store/yw8yb6x8iad55hnwr77ch198864qbjrw-python3.10-ipyvolume-0.5.2/lib/python3.10/site-packages/ipyvolume/styles.py", line 56, in <module>
       >     utils.dict_deep_update(default, _defaults)
       >   File "/nix/store/yw8yb6x8iad55hnwr77ch198864qbjrw-python3.10-ipyvolume-0.5.2/lib/python3.10/site-packages/ipyvolume/utils.py", line 19, in dict_deep_update
       >     if isinstance(v, collections.Mapping):
       > AttributeError: module 'collections' has no attribute 'Mapping'
       For full logs, run 'nix log /nix/store/j4sahn2primmfn8plqy50jc48knnbk63-python3.10-ipyvolume-0.5.2.drv'.
error: 1 dependencies of derivation '/nix/store/sds3d5gydk8afm23b0rgyx0f3bsv5ycl-python3.10-niwidgets-0.2.2.drv' failed to build
error: builder for '/nix/store/b4p1afagb34pb2xqs1hgw3jhhgfpa2d6-python3.9-ipyvolume-0.5.2.drv' failed with exit code 1;
       last 10 log lines:
       >     from . import datasets
       >   File "/nix/store/qp7wchxnk5rirfn0vi3qpy8d75iry0q8-python3.9-ipyvolume-0.5.2/lib/python3.9/site-packages/ipyvolume/datasets.py", line 8, in <module>
       >     os.makedirs(data_dir)
       >   File "/nix/store/bvzpppawf2naqv3qhxqv66jxaq3iaxyj-python3-3.9.11/lib/python3.9/os.py", line 215, in makedirs
       >     makedirs(head, exist_ok=exist_ok)
       >   File "/nix/store/bvzpppawf2naqv3qhxqv66jxaq3iaxyj-python3-3.9.11/lib/python3.9/os.py", line 215, in makedirs
       >     makedirs(head, exist_ok=exist_ok)
       >   File "/nix/store/bvzpppawf2naqv3qhxqv66jxaq3iaxyj-python3-3.9.11/lib/python3.9/os.py", line 225, in makedirs
       >     mkdir(name, mode)
       > PermissionError: [Errno 13] Permission denied: '/homeless-shelter'
       For full logs, run 'nix log /nix/store/b4p1afagb34pb2xqs1hgw3jhhgfpa2d6-python3.9-ipyvolume-0.5.2.drv'.
error: 1 dependencies of derivation '/nix/store/f0w74rqqv96izgsyc6l6pb4zf3qawp4f-python3.9-niwidgets-0.2.2.drv' failed to build
error: 4 dependencies of derivation '/nix/store/dphl08azjvwc1550f7ynqimv5livc801-review-shell.drv' failed to build

Link to currently reviewing PR:
https://github.com/NixOS/nixpkgs/pull/168423

4 packages failed to build:
python310Packages.ipyvolume python310Packages.niwidgets python39Packages.ipyvolume python39Packages.niwidgets

4 packages built:
python310Packages.ipywebrtc python310Packages.pythreejs python39Packages.ipywebrtc python39Packages.pythreejs
```

**Comment by cfhammill** [comment]
Thanks @bcdarwin, I've been working in python39, haven't checked against 310 yet, good to know there are some incompabilities there. Is failing under 310 a blocker to merge, or can we flag that this doesn't work on 310 yet?

> The `pythonImportsCheck` for `niwidgets` appears to be failing; maybe you need to `export HOME=$TMPDIR` or similar for that phase?

What error are you getting and is this with python3Packages, or a specific version? It's working for me but as I'm not on nixOS that doesn't mean much.


> Stylistically, a few more blank lines between logical sections of the expression would be slightly nicer.

What do you consider to be the logical sections? package level vars + src fetch, dependendencies, and meta maybe?

**Comment by cfhammill** [comment]
Also looking at the OfBorg failures for darwin and linux, it looks like the the test directory is read only during check phase where the test suite is attempting to create files. Any suggestions for how to work around this (it doesn't seem to be the case when I build locally, don't know if this is CI or NixOS or something weird about my set up)

**Comment by cfhammill** [comment]
Build failures under 310 seem to be the use of `collections.Mapping`, which is now `collections.abc.Mapping`

**Comment by bcdarwin** [comment]
> Also looking at the OfBorg failures for darwin and linux, it looks like the the test directory is read only during check phase where the test suite is attempting to create files. Any suggestions for how to work around this (it doesn't seem to be the case when I build locally, don't know if this is CI or NixOS or something weird about my set up)

See my above comment about setting `$HOME`.  Not sure why the tests pass on your machine, though - they should not (irrespective of NixOS or not).

**Comment by cfhammill** [comment]
> > Also looking at the OfBorg failures for darwin and linux, it looks like the the test directory is read only during check phase where the test suite is attempting to create files. Any suggestions for how to work around this (it doesn't seem to be the case when I build locally, don't know if this is CI or NixOS or something weird about my set up)
> 
> See my above comment about setting `$HOME`. Not sure why the tests pass on your machine, though - they should not (irrespective of NixOS or not).

If the issue is $HOME is unset and the tests are trying to write there, It's possible I forgot to build purely at some point and have been hitting the cached version. This seems like this has nothing to do with `pythonImportsCheck` though, should break the normal check phase though. 

> > Build failures under 310 seem to be the use of `collections.Mapping`, which is now `collections.abc.Mapping`
> > This has been fixed in the github version but hasn't made it to pypi yet (currently release was 0.5.2 from 2019, github is tagged at 0.6.0-alpha10), happy to migrate to the github version if this is a blocker.
> 
> You should set `disabled = pythonOlder "..."` in general, and for this package you can either update to the GitHub version or set `meta.broken = ...`.

It would be disabling pythonNewer I think, meta.broken would be annoying because the package isn't broken, it just isn't compatible with 310 (yet). I think switching to the github version should be ok - I'll give that a go.

**Comment by bcdarwin** [comment]
> > > Also looking at the OfBorg failures for darwin and linux, it looks like the the test directory is read only during check phase where the test suite is attempting to create files. Any suggestions for how to work around this (it doesn't seem to be the case when I build locally, don't know if this is CI or NixOS or something weird about my set up)
> > 
> > 
> > See my above comment about setting `$HOME`. Not sure why the tests pass on your machine, though - they should not (irrespective of NixOS or not).
> 
> If the issue is $HOME is unset and the tests are trying to write there, It's possible I forgot to build purely at some point and have been hitting the cached version. This seems like this has nothing to do with `pythonImportsCheck` though, should break the normal check phase though.
> 
> > > Build failures under 310 seem to be the use of `collections.Mapping`, which is now `collections.abc.Mapping`
> > > This has been fixed in the github version but hasn't made it to pypi yet (currently release was 0.5.2 from 2019, github is tagged at 0.6.0-alpha10), happy to migrate to the github version if this is a blocker.
> > 
> > 
> > You should set `disabled = pythonOlder "..."` in general, and for this package you can either update to the GitHub version or set `meta.broken = ...`.
> 
> It would be disabling pythonNewer I think, meta.broken would be annoying because the package isn't broken, it just isn't compatible with 310 (yet). I think switching to the github version should be ok - I'll give that a go.

Sorry, what I meant was that you should additionally add the lower bound for these packages (even if they're lower than what Hydra builds).

Switching to the Github version is probably fine but I'd be happy to have the non-alpha version with `meta.broken` set -- it's a fairly common idiom in `python-modules`.

**Comment by bcdarwin** [comment]
Closing as `niwidgets` appears to be abandoned and unbuildable with recent Python packages (`ipywidgets` etc.).

Of course, it may still be worthwhile to add updated versions of the dependencies included in this PR at some point.

**Pull request closed without merging**